File: backend/widgets/src/main/kotlin/br/com/zup/beagle/widget/form/SimpleForm.kt
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package br.com.zup.beagle.widget.form

import br.com.zup.beagle.builder.BeagleListBuilder
import br.com.zup.beagle.core.ServerDrivenComponent
import br.com.zup.beagle.widget.action.Action
import br.com.zup.beagle.widget.builder.BeagleWidgetBuilder
import br.com.zup.beagle.widget.context.ContextData
import kotlin.properties.Delegates

/**
 * Component will define a submit handler for a SimpleForm.
 *
 * @param context define the contextData that be set to form
 *
 * @param children define the items on the simple form.
 *
 * @param onSubmit define the actions you want to execute when action submit form
 *
 */
class SimpleForm(
    val context: ContextData,
    val onSubmit: List<Action>,
    val children: List<ServerDrivenComponent>
) : ServerDrivenComponent {
    companion object{
        @JvmStatic
        fun builder() = Builder()
    }
    class Builder : BeagleWidgetBuilder<SimpleForm> {
        var context: ContextData by Delegates.notNull()
        var onSubmit: MutableList<Action> by Delegates.notNull()
        var children: MutableList<ServerDrivenComponent> by Delegates.notNull()

        fun context(context: ContextData) = this.apply { this.context = context }

        fun onSubmit(onSubmit: List<Action>) = this.apply { this.onSubmit = onSubmit.toMutableList() }

        fun children(children: List<ServerDrivenComponent>)
            = this.apply { this.children = children.toMutableList() }

        fun context(block: ContextData.Builder.() -> Unit) {
            context(ContextData.Builder().apply(block).build())
        }

        fun onSubmit(block: BeagleListBuilder<Action>.() -> Unit) {
            onSubmit(BeagleListBuilder<Action>().apply(block).build())
        }

        fun children(block: BeagleListBuilder<ServerDrivenComponent>.() -> Unit) {
            children(BeagleListBuilder<ServerDrivenComponent>().apply(block).build())
        }

        override fun build() = SimpleForm(
            context = context,
            onSubmit = onSubmit,
            children = children
        )
    }
}

fun simpleForm(block: SimpleForm.Builder.() -> Unit) = SimpleForm.Builder().apply(block).build()

File: iOS/Sources/Beagle/Sources/Components/Widget/SimpleForm/SimpleForm.swift
/*
 * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/// Component will define a submit handler for a `SimpleForm`.
public struct SimpleForm: ServerDrivenComponent, HasContext, AutoInitiableAndDecodable {
    
    /// Defines the contextData that be set to form.
    public var context: Context?
    
    /// Defines the actions you want to execute when action submit form.
    public let onSubmit: [Action]?
    
    /// This event is executed every time a form is submitted, but because of a validation error, the onSubmit event is not run.
    public let onValidationError: [Action]?
    
    /// Defines the items on the simple form.
    public let children: [ServerDrivenComponent]?
    
    /// Properties that all widgets have in common.
    public var widgetProperties: WidgetProperties
    
// sourcery:inline:auto:SimpleForm.Init
    public init(
        context: Context? = nil,
        onSubmit: [Action]? = nil,
        onValidationError: [Action]? = nil,
        children: [ServerDrivenComponent]? = nil,
        widgetProperties: WidgetProperties = WidgetProperties()
    ) {
        self.context = context
        self.onSubmit = onSubmit
        self.onValidationError = onValidationError
        self.children = children
        self.widgetProperties = widgetProperties
    }
// sourcery:end
    
    public init(
        context: Context? = nil,
        onSubmit: [Action]? = nil,
        onValidationError: [Action]? = nil,
        widgetProperties: WidgetProperties = WidgetProperties(),
        @ChildrenBuilder
        _ children: () -> [ServerDrivenComponent]
    ) {
        self.init(context: context, onSubmit: onSubmit, onValidationError: onValidationError, children: children(), widgetProperties: widgetProperties)
    }
    
    #if swift(<5.3)
    public init(
        context: Context? = nil,
        onSubmit: [Action]? = nil,
        onValidationError: [Action]? = nil,
        widgetProperties: WidgetProperties = WidgetProperties(),
        @ChildBuilder
        _ children: () -> ServerDrivenComponent
    ) {
        self.init(context: context, onSubmit: onSubmit, onValidationError: onValidationError, children: [children()], widgetProperties: widgetProperties)
    }
    #endif
}
